dm/master/agent_pool.go

```diff
@@ -36,42 +52,62 @@ type Agent struct {
 }
 
 // NewAgentPool returns a agent pool
-func NewAgentPool(limit int) *AgentPool {
-	agents := make(chan *Agent, limit)
-	for i := 0; i < limit; i++ {
-		agents <- &Agent{ID: i + 1}
-	}
+func NewAgentPool(cfg *RateLimitConfig) *AgentPool {
+	requests := make(chan int, int(math.Ceil(1/cfg.rate))+cfg.burst)
+	agents := make(chan *Agent, cfg.burst)
+	limiter := rate.NewLimiter(rate.Limit(cfg.rate), cfg.burst)
 
 	return &AgentPool{
-		limit:  limit,
-		agents: agents,
+		requests: requests,
+		agents:   agents,
+		cfg:      cfg,
+		limiter:  limiter,
 	}
 }
 
 // Apply applies for a agent
-func (pool *AgentPool) Apply() *Agent {
-	agent := <-pool.agents
-	return agent
-}
+// if ctx is canceled before we get an agent, returns nil
+func (ap *AgentPool) Apply(ctx context.Context, id int) *Agent {
+	select {
+	case <-ctx.Done():
+		return nil
+	case ap.requests <- id:
+	}
 
-// Recycle recycles agent
-func (pool *AgentPool) Recycle(agent *Agent) {
-	pool.agents <- agent
+	select {
+	case <-ctx.Done():
+		return nil
+	case agent := <-ap.agents:
+		return agent
+	}
 }
 
-// GetAgentPool a singleton agent pool
-func GetAgentPool() *AgentPool {
-	once.Do(func() {
-		pool = NewAgentPool(agentlimit)
-	})
-	return pool
+// Start starts AgentPool background dispatcher
+func (ap *AgentPool) Start(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case id := <-ap.requests:
+			err := ap.limiter.Wait(ctx)
+			if err == context.Canceled {
```

Comment:
what if other non-nil error?

Reply:
other error should not happen, panic if meets other error here